exporter/lokiexporter/exporter.go

```diff
@@ -0,0 +1,195 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lokiexporter
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.uber.org/zap"
+)
+
+type lokiExporter struct {
+	config             *Config
+	logger             *zap.Logger
+	client             *http.Client
+	attributesToLabels map[string]model.LabelName
+	wg                 sync.WaitGroup
+}
+
+func newExporter(config *Config, logger *zap.Logger) (*lokiExporter, error) {
+	client, err := config.HTTPClientSettings.ToClient()
+	if err != nil {
+		return nil, err
+	}
+
+	return &lokiExporter{
+		config: config,
+		logger: logger,
+		client: client,
+	}, nil
+}
+
+func (l *lokiExporter) pushLogData(ctx context.Context, ld pdata.Logs) (numDroppedLogs int, err error) {
+	l.wg.Add(1)
+	defer l.wg.Done()
+
+	pushReq, numDroppedLogs := l.logDataToLoki(ld)
+	if len(pushReq.Streams) == 0 {
+		return numDroppedLogs, nil
+	}
+
+	buf, err := encode(pushReq)
+	if err != nil {
+		numDroppedLogs += ld.LogRecordCount()
+		return numDroppedLogs, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", l.config.HTTPClientSettings.Endpoint, bytes.NewReader(buf))
+	if err != nil {
+		return numDroppedLogs, consumererror.Permanent(err)
+	}
+
+	for k, v := range l.config.HTTPClientSettings.Headers {
+		req.Header.Set(k, v)
+	}
+	req.Header.Set("Content-Type", "application/x-protobuf")
+
+	if len(l.config.TenantID) > 0 {
+		req.Header.Set("X-Scope-OrgID", l.config.TenantID)
+	}
+
+	resp, err := l.client.Do(req)
+	if err != nil {
+		numDroppedLogs += ld.LogRecordCount()
+		return numDroppedLogs, err
+	}
+
+	_, _ = io.Copy(ioutil.Discard, resp.Body)
+	_ = resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
```

Comment:
Retries and other resiliency handling is expected to be implemented in an exporter. exporterhelper package in opentelemetry-collector project has functionality for assisting in this.

Reply:
Great catch, @kbrockhoff! I forgot to add those back in.

Reply:
Updated with the appropriate consumererrors so that the export helper would retry when appropriate.